src/stripeLogsView.ts

```diff
@@ -1,15 +1,203 @@
+  private logsStderrStream: stream.Writable | null;
+
+  constructor(stripeClient: StripeClient) {
+    super();
+    this.stripeClient = stripeClient;
+    this.logTreeItems = [];
+    this.logsStdoutStream = null;
+    this.logsStderrStream = null;
+    this.viewState = ViewState.Idle;
+  }
+
+  async startLogsStreaming() {
+    this.setViewState(ViewState.Loading);
+    try {
+      await this.setupStreams();
+      this.setViewState(ViewState.Streaming);
+    } catch (e) {
+      window.showErrorMessage(e.message);
+      this.cleanupStreams();
+      this.setViewState(ViewState.Idle);
+    }
+  }
+
+  stopLogsStreaming() {
+    this.cleanupStreams();
+    this.setViewState(ViewState.Idle);
+  }
+
   buildTree(): Promise<StripeTreeItem[]> {
-    const logStreamItem = new StripeTreeItem('Start API logs streaming', 'openLogsStreaming');
-    logStreamItem.setIcon({
-      dark: Resource.ICONS.dark.terminal,
-      light: Resource.ICONS.light.terminal,
-    });
+    const streamingControlItemArgs = (() => {
+      switch (this.viewState) {
+        case ViewState.Idle:
+          return {
+            label: 'Start streaming API logs',
+            command: 'startLogsStreaming',
+            iconId: 'play-circle',
+          };
+        case ViewState.Loading:
+          return {
+            label: 'Starting streaming API logs...',
+            command: 'stopLogsStreaming',
+            iconId: 'loading',
+          };
+        case ViewState.Streaming:
+          return {
+            label: 'Stop streaming API logs',
+            command: 'stopLogsStreaming',
+            iconId: 'stop-circle',
+          };
+      }
+    })();
+
+    const streamingControlItem = this.createItemWithCommand(streamingControlItemArgs);
+
+    const treeItems = [streamingControlItem];
+
+    if (this.logTreeItems.length > 0) {
+      const logsStreamRootItem = new StripeTreeItem('Recent logs');
+      logsStreamRootItem.children = this.logTreeItems;
+      logsStreamRootItem.expand();
+      treeItems.push(logsStreamRootItem);
+    }
+
+    return Promise.resolve(treeItems);
+  }
+
+  private createItemWithCommand({
+    label,
+    command,
+    iconId,
+  }: {
+    label: string;
+    command?: string;
+    iconId?: string;
+  }) {
+    const item = new StripeTreeItem(label, command);
+    if (iconId) {
+      item.iconPath = new ThemeIcon(iconId);
+    }
+    return item;
+  }
+
+  private async setupStreams() {
+    const stripeLogsTailProcess = await this.stripeClient.getOrCreateCLIProcess(
+      CLICommand.LogsTail,
+      ['--format', 'JSON'],
+    );
+    if (!stripeLogsTailProcess) {
+      throw new Error('Failed to start `stripe logs tail` process');
+    }
+
+    stripeLogsTailProcess.on('close', this.cleanupStreams);
+    stripeLogsTailProcess.on('error', this.cleanupStreams);
+
+    // The CLI lets you know that streaming is ready via stderr
+    if (!this.logsStderrStream) {
+      await new Promise<void>((resolve) => {
```

Comment:
what happens if the process fails to invoke while we're waiting on this promise? Seems like we'll be notified in the `error` handler for ` stripeLogsTailProcess.on('error'` which will trigger cleaning up the streams (which haven't entirely been completed yet, but we'll never finish this?

Relatedly, what if it's taking a "long" time for the cli to report itself as being ready, will we have a deadline here?

Reply:
> what happens if the process fails to invoke while we're waiting on this promise? Seems like we'll be notified in the error handler for stripeLogsTailProcess.on('error' which will trigger cleaning up the streams (which haven't entirely been completed yet, but we'll never finish this?

If the process fails to invoke, then this promise will never resolve and the UI will stay in a loading state until the user cancels the command. So naively, I'd say this promise needs to reject when the process has an `error`. To implement that, maybe reject the promise in the handler of `stripeLogsTailProcess.on('error'`? This would mean moving this Promise instantiation one level out so that `reject` is in scope of the error handler. Would this be reasonable, or is another approach needed?

> Relatedly, what if it's taking a "long" time for the cli to report itself as being ready, will we have a deadline here?

Assuming the process doesn't error out, I can think of these scenarios:
- The CLI is getting ready, but it hasn't reported anything after a long time
  - A deadline might be appropriate here. What should we do if we reach the deadline and the CLI isn't ready? I think we should notify the user that it's taking a long time and have them decide if they want to cancel it or not. The UI can support cancelling the command while loading.
- The CLI reports some status other than "ready"
  - In this case, we should show the status to the user, stop the command and clean up.

Reply:
Right, all 3 cases (failed to invoke, deadline exceeded, and unexpected status message) feel like shades of the same question of what the general approach is for handling process failures. Much of that boils down to, what are we going to show the user and when? And how proactive do we need to be in taking action?

I'm not sure I have a strong opinion on how to handle the ordering of the promise and the error handling, other than asking myself does it actually matter that we need to wait to see the `Ready!` message before configuring the `stdout` handler? There's no strict dependency between these pieces, we could just blindly attach the handlers and move on and let the basic error handling take care of cleaning things up?

Reply:
- I think that's an accurate and succinct summary. I'll be coming back to this as I'm working to make sure I'm thinking about this correctly.
- That's a good point, a promise might not actually be needed here! Would simplify things if it can be removed.